### PR TITLE
Make builds less verbose

### PR DESCRIPTION
**Pull request by daverodgman**
## Description

Experiment with wrapping make and cmake to hide output (when there are no errors), to reduce logging load on Jenkins.

Fixes https://github.com/Mbed-TLS/mbedtls-test/issues/149

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required
- [ ] **backport** #8873 
- [x] **tests** not required

### CODE REVIEW

File: tests/scripts/quiet/make [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/make:1]: Note that on FreeBSD, [the Groovy code](https://github.com/Mbed-TLS/mbedtls-test/blob/865d01b47c39e6f9da39683ee2a2105fee767797/vars/gen_jobs.groovy#L119) sets up a wrapper to arrange for `make` to call GNU make (on FreeBSD, `make` is BSD make and our makefiles are not compatible with BSD make). The two wrappers may interfere with each other. (I haven't analyzed your code to see if they do interfere, just pointing out the risk.)

You have my blessing if you want to update the build scripts in the mbedtls repository to not need the hack anymore. I didn't do it at the time partly because we had to make FreeBSD work on unpatched branches, and partly because there's a nontrivial amount of places where we'd need to customize the `make` invocation. Note that even if you do that we'll need to keep the mbedtls-test hack for a while, for the sake of branches that don't have the improvement.

Comment (daverodgman) [reply tests/scripts/quiet/make]: This turns out not to have any impact on this script (it just calls the next thing called "make" on the path, which does the right thing in the presence of the FreeBSD wrapper). So I won't touch this for now.

File: tests/scripts/quiet/make [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+# Locate original tool
+ORIGINAL_TOOL=$(type -ap ${TOOL} | grep -v "$0" | head -n1 )
+
+if [[ " $@ " =~ $NO_SILENCE || -n "${VERBOSE_LOGS}" ]]; then
+    ${ORIGINAL_TOOL} "$@"
+    EXIT_STATUS=$?
+else
+    # Display the command being invoked - if it succeeds, this is all that will
+    # be displayed.
+    echo "${TOOL} $@"
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/make:27]: Displaying the `make` invocation is something I'd like to do in all cases. It would be very helpful to re-create complex builds, such as builds with drivers, with tweaked arguments (typically enabling debug). Since you've already set up a `make` wrapper, please:

* Display this regardless of the verbosity setting.
* Use an unambiguous way to display the command line: it's vital to distinguish spaces inside arguments and spaces that separate arguments. `printf '%q ' "$@"; printf '\n'` is a bit ugly but works. (At least with bash 4, I need to confirm if macOS's antique bash already supports `%q`.)

File: tests/scripts/quiet/cmake [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    echo "${TOOL} $(quote_args "$@")"
+fi
+
+if [[ " $@ " =~ $NO_SILENCE || -n "${VERBOSE_LOGS}" ]]; then
+    # Run original command with no output supression
+    ${ORIGINAL_TOOL} "$@"
+    EXIT_STATUS=$?
+else
+    # Run original command and capture output & exit status
+    TMPFILE=$(mktemp /tmp/quiet-${TOOL}.XXXXXX)
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/cmake:55]: Let `TMPDIR` override the default choice of `/tmp` by using a relative path.

```suggestion
    TMPFILE=$(mktemp quiet-${TOOL}.XXXXXX)
```

File: tests/scripts/quiet/cmake [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        else
+            # get bash to do the quoting
+            q=$(printf '%q' "$a")
+        fi
+        s="$s $q"
+    done
+    echo $s
+}
+
+if [[ ! " $@ " =~ " --version " ]]; then
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/cmake:42]: Minor: `" $@ "` is correct but weird: the difference between `$*` and `$@` is that `"<PREFIX>$@<SUFFIX>"` is multiple words when used in list context, but here the string is in word context so the result is the same as `" $* "`: concatenation of the arguments with a space in between. Since the concatenation is intended, it's less surprising to write `" $* "`.

File: tests/scripts/quiet/cmake [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+# by un-commenting the following line (or setting VERBOSE_LOGS in your environment):
+# VERBOSE_LOGS=1
+
+# don't silence invocations containing these arguments
+NO_SILENCE=" --version "
+
+TOOL=$(basename "$0")
+
+# Locate original tool
+ORIGINAL_TOOL=$(type -ap ${TOOL} | grep -v "$0" | head -n1 )
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/cmake:19]: Avoid corner cases with a path containing wildcard characters or being a suffix of another:
```suggestion
ORIGINAL_TOOL=$(type -ap "${TOOL}" | grep -v -Fx "$0" | head -n1 )
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Shell quoting is hard.

File: tests/scripts/quiet/cmake [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+quote_args() {
```
[the remark this answers is not in the document]
Comment (tom-cosgrove-arm) [reply tests/scripts/quiet/cmake:21]: Do we need to be zsh-safe as `all.sh` starts with `#! /usr/bin/env bash`? You'd actually have to run it explicitly with `zsh all.sh` to have `zsh` running as the interpreter. (To be clear, I'm a `ksh` user :) and I just run `all.sh` with `tests/scripts/all.sh`. It doesn't look like `all.sh` supports `ksh` (`ksh` doesn't have `shopt`), so why should it support `zsh`?

File: tests/scripts/quiet/cmake [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    # similar to printf '%q' "$@"
+    # but produce more human-readable results for common/simple cases like "a b"
+    local args=("$@")
+    s=""
+    for a in "${args[@]}"; do
+        simple_pattern='^[[:alnum:] _=+-]*$'
```
[the remark this answers is not in the document]
Comment (tom-cosgrove-arm) [reply tests/scripts/quiet/cmake:27]: Do we need to be zsh-safe as `all.sh` starts with `#! /usr/bin/env bash`? You'd actually have to run it explicitly with `zsh all.sh` to have `zsh` running as the interpreter. (To be clear, I'm a `ksh` user :) and I just run `all.sh` with `tests/scripts/all.sh`). It doesn't look like `all.sh` supports `ksh` (`ksh` doesn't have `shopt`), so why should it support `zsh`?

File: tests/scripts/quiet/cmake (hunk 6 above)
[the remark this answers is not in the document]
Comment (tom-cosgrove-arm) [reply tests/scripts/quiet/cmake:21]: Do we need to be zsh-safe as `all.sh` starts with `#! /usr/bin/env bash`? You'd actually have to run it explicitly with `zsh all.sh` to have `zsh` running as the interpreter. (To be clear, I'm a `ksh` user :) and I just run `all.sh` with `tests/scripts/all.sh`). It doesn't look like `all.sh` supports `ksh` (`ksh` doesn't have `shopt`), so why should it support `zsh`?

(For some reason Chrome ended up in wrong text box)

[the remark this answers is not in the document]
Comment (tom-cosgrove-arm) [reply tests/scripts/quiet/cmake:21]: Grr, I managed to delete @gilles-peskine-arm comment by mistake, which said

> This is both simpler and, I think, produces nicer output. (But doesn't make the output fully zsh-safe in unlikely cases involving equal signs that I mention in another comment.)
>
> ```
print_quoted_args () {
    local a
    for a in "$@"; do
        if [[ $a == *[!-%+,./:=[:alnum:]@_]* || $a == '' ]]; then
            # Wrap in single quotes. Single quotes in the string need to be transformed.
            a="'${a//\'/\'\\\'\'}'"
        fi
    printf "%s " "$a"
    done
    echo
}
```

File: tests/scripts/quiet/cmake [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        else
+            # get bash to do the quoting (which may result in no quotes or escaping,
+            # if none is needed).
+            q=$(printf '%q' "$a")
+        fi
+        printf "%s " "$q"
+    done
+}
+
+if [[ ! " $* " =~ " --version " ]]; then
```
Comment (tom-cosgrove-arm) [inline tests/scripts/quiet/cmake:42]: Why do we hardcode `" --version "` here when we have `NO_SILENCE`?

File: tests/scripts/quiet/cmake [hunk 9]
```diff
@@ -0,0 +1,70 @@
+#! /usr/bin/env bash
```
Comment (tom-cosgrove-arm) [inline tests/scripts/quiet/cmake:1]: Since the contents of `quiet/cmake` and `quiet/make` are the same except for the value of `NO_SILENCE`, is there any way to extract the common parts?

File: tests/scripts/quiet/cmake (hunk 8 above)
Comment (daverodgman) [reply tests/scripts/quiet/cmake:42]: This is different in `make`.

For commands with `--version`, we always avoid printing the command being executed to make sure the output doesn't break scripts that parse it.

For NO_SILENCE, we might want to print the command being executed, as well as its output (as with `make test`).

File: tests/scripts/quiet/cmake (hunk 9 above)
Comment (tom-cosgrove-arm) [reply tests/scripts/quiet/cmake:1]: That is much nicer

File: tests/scripts/quiet/quiet [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+# Locate original tool
+TOOL_WITH_PATH=$(dirname "$0")/$TOOL
+ORIGINAL_TOOL=$(type -ap "${TOOL}" | grep -v -Fx "$TOOL_WITH_PATH" | head -n1)
+
+print_quoted_args() {
+    # similar to printf '%q' "$@"
+    # but produce more human-readable results for common/simple cases like "a b"
+    for a in "$@"; do
+        simple_pattern='^([[:alnum:]_+-]+=)?([[:alnum:] _=+-./:@]*)$'
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet:21]: `+-./` in the character set looks like a gotcha. It's the range from `+` to `.` and the character `/`. The range from `+` to `.` is the four characters `+,-.` — the set includes the comma but you might miss it at first glance. If you're going to use ranges, you might as well use `+-/` since `/` follows `.`. However, all of these observations about ranges are true only if the ASCII punctuation symbols are in order and with no other character interspersed. I don't know if that's the case in practice. It would be safer to avoid ranges. A hyphen does not denote a range if it's immediately after the opening `]` or immediately before the closing `]`.
```suggestion
        simple_pattern='^([[:alnum:]_+-]+=)?([-[:alnum:] _=+./:@]*)$'
```

File: tests/scripts/quiet/quiet [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        if [[ $a =~ ' ' && $a =~ $simple_pattern ]]; then
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet:22]: Why not use single quotes? Then the only character you'd have to worry about is `'` itself.
```suggestion
        if [[ $a =~ ' ' && $a =~ ^([-+./:_[:alnum:]]=)?[^\']+$ ]]; then
```

Also, how about triggering this path with other characters that need quoting, e.g. for `MBEDTLS_CONFIG_FILE="my/config.h"`?
```suggestion
        if [[ $a =~ [^-+.,/:=@[:alnum:]_] && $a =~ ^([^-+.,/:@[:alnum:]_]=)?[^\']+$ ]]; then
```

File: tests/scripts/quiet/quiet [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    done
+}
+
+if [[ ! " $* " =~ " --version " ]]; then
+    # Display the command being invoked - if it succeeds, this is all that will
+    # be displayed. Don't do this for invocations with --version, because
+    # this output is often parsed by scripts, so we don't want to modify it.
+    printf %s "${TOOL} "
+    print_quoted_args "$@"
+    echo
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet:44]: Maybe this should go to stderr, to minimize the risk of disruption?

(But still quiet it for `--version`.)

File: tests/scripts/quiet/quiet [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        printf "%s " "$q"
+    done
+}
+
+if [[ ! " $* " =~ " --version " ]]; then
+    # Display the command being invoked - if it succeeds, this is all that will
+    # be displayed. Don't do this for invocations with --version, because
+    # this output is often parsed by scripts, so we don't want to modify it.
+    printf %s "${TOOL} "
+    print_quoted_args "$@"
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet:43]: Minor: we could remove the space after `${TOOL}` and have `print_quoted_args` print a space before each argument, not after. This would avoid the trailing space which is generally invisible but can come up in trailing space and can cause an extra visual line if it happens to occur at the right margin.

File: tests/scripts/quiet/quiet [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+# Copyright The Mbed TLS Contributors
+# SPDX-License-Identifier: Apache-2.0 OR GPL-2.0-or-later
+#
+# This swallows the output of the wrapped tool, unless there is an error.
+# This helps reduce excess logging in the CI.
+
+# If you are debugging a build / CI issue, you can get complete unsilenced logs
+# by un-commenting the following line (or setting VERBOSE_LOGS in your environment):
+# VERBOSE_LOGS=1
+
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet:12]: Please document that this script expects its caller to set `NO_SILENCE` and `TOOL`, with a short explanation of their meaning.

File: tests/scripts/quiet/quiet [hunk 15]
```diff
@@ -0,0 +1,66 @@
+#! /usr/bin/env bash
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet:1]: Why is this executable, rather than a shell library sourced by the others?

So: no shebang in `quiet`, it's not executable, and each caller is
```
#!/usr/bin/env bash
NO_SILENCE=…
TOOL=…
"$(dirname "$0")/quiet.bash" "$@"
```

File: tests/scripts/quiet/make [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+# If you are debugging a build / CI issue, you can get complete unsilenced logs
+# by un-commenting the following line (or setting VERBOSE_LOGS in your environment):
+# VERBOSE_LOGS=1
+
+# don't silence invocations containing these arguments
+export NO_SILENCE=" --version | test "
+
+export TOOL="make"
+
+exec $(dirname "$0")/quiet "$@"
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/make:18]:
```suggestion
exec "$(dirname "$0")/quiet" "$@"
```
in case the directory name contains special characters.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
This mostly looks good now, but I'd like some improvements to the quoting and some simplifications.

File: tests/scripts/quiet/quiet [hunk 17]
```diff
@@ -9,6 +9,17 @@
...
+#
+# This script provides most of the functionality for the adjacent make and cmake
+# wrappers.
+#
+# It requires two variables to be set:
+#
+# TOOL       - the name of the tool that is being wrapped (with no path), e.g. "make"
+#
+# NO_SILENCE - a regex that describes the commandline arguments for which output will not
+#              be silenced, e.g. " --version | test ". In this example, "make test" will
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet:19]: Minor: `make lib test` would be a better example, to illustrate that the regex doesn't need to match all the arguments.

File: tests/scripts/quiet/quiet [hunk 18]
```diff
@@ -27,18 +27,15 @@ print_quoted_args() {
...
-            # simplify quoted form - e.g.:
-            #   a b        -> "a b"
-            #   CFLAGS=a b -> CFLAGS="a b"
-            q="${BASH_REMATCH[1]}\"${BASH_REMATCH[2]}\""
-        else
-            # get bash to do the quoting (which may result in no quotes or escaping,
-            # if none is needed).
-            q=$(printf '%q' "$a")
+        # Get bash to quote the string
+        q=$(printf '%q' "$a")
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet:31]: Minor performance efficiency gain:
```
printf -v q '%q' "$a"

File: tests/scripts/quiet/quiet [hunk 19]
```diff
@@ -1,5 +1,3 @@
-#! /usr/bin/env bash
-#
 # Copyright The Mbed TLS Contributors
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet:1]: By our naming conventions, this file should be called `quiet.sh` (or `quiet.bash`). Not a blocker as far as I'm concerned.

Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet:1]: It would be nice to have some indication that this script uses bash. For example you can declare it to Emacs by putting this on the first line:
```
# -*-mode: sh; sh-shell: bash -*-
```
(Would be nice to have the Vim equivalent but I don't know what it is.)

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM, thanks for keeping up with the iterations. I left a few suggestions but none are blockers.

File: tests/scripts/quiet/quiet.sh [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+# This script provides most of the functionality for the adjacent make and cmake
+# wrappers.
+#
+# It requires two variables to be set:
+#
+# TOOL       - the name of the tool that is being wrapped (with no path), e.g. "make"
+#
+# NO_SILENCE - a regex that describes the commandline arguments for which output will not
+#              be silenced, e.g. " --version | test ". In this example, "make test" will
+#              not be silent, but "make lib test" will be.
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet.sh:22]:
```suggestion
#              be silenced, e.g. " --version | test ". In this example, "make lib test" will
#              not be silent, but "make lib" will be.
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM apart from a comment that's the wrong way round

**Review by gilles-peskine-arm — DISMISSED** [review]
LGTM

File: tests/scripts/quiet/cmake [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+#
+# Copyright The Mbed TLS Contributors
+# SPDX-License-Identifier: Apache-2.0 OR GPL-2.0-or-later
+#
+# This swallows the output of the wrapped tool, unless there is an error.
+# This helps reduce excess logging in the CI.
+
+# If you are debugging a build / CI issue, you can get complete unsilenced logs
+# by un-commenting the following line (or setting VERBOSE_LOGS in your environment):
+# VERBOSE_LOGS=1
```
Comment (tom-cosgrove-arm) [inline tests/scripts/quiet/cmake:11]: Can you actually get unsilenced logs by un-commenting this line?

File: tests/scripts/quiet/make [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+#
+# Copyright The Mbed TLS Contributors
+# SPDX-License-Identifier: Apache-2.0 OR GPL-2.0-or-later
+#
+# This swallows the output of the wrapped tool, unless there is an error.
+# This helps reduce excess logging in the CI.
+
+# If you are debugging a build / CI issue, you can get complete unsilenced logs
+# by un-commenting the following line (or setting VERBOSE_LOGS in your environment):
+# VERBOSE_LOGS=1
```
Comment (tom-cosgrove-arm) [inline tests/scripts/quiet/make:11]: Can you actually get unsilenced logs by un-commenting this line?

File: tests/scripts/quiet/quiet.sh [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+#
+# Copyright The Mbed TLS Contributors
+# SPDX-License-Identifier: Apache-2.0 OR GPL-2.0-or-later
+#
+# This swallows the output of the wrapped tool, unless there is an error.
+# This helps reduce excess logging in the CI.
+
+# If you are debugging a build / CI issue, you can get complete unsilenced logs
+# by un-commenting the following line (or setting VERBOSE_LOGS in your environment):
+# VERBOSE_LOGS=1
```
Comment (tom-cosgrove-arm) [inline tests/scripts/quiet/quiet.sh:11]: It's hard to notice this paragraph. How about either a blank line instead of line 12, or putting this after the description of `NO_SILENE`?

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

File: tests/scripts/quiet/quiet.sh [hunk 24]
```diff
@@ -0,0 +1,75 @@
+# -*-mode: sh; sh-shell: bash -*-
```
Comment (gilles-peskine-arm) [inline tests/scripts/quiet/quiet.sh:1]: Sorry for not noticing this earlier but there's an inconsistency:

* If `quiet.sh` is executable, it should have a shebang.
* If `quiet.sh` is not executable, it should not have the execute flag and it should be included with `.` rather than executed.

The current code works thanks to bash's fallback behavior: `execve("path/to/quiet.sh", ...)` returns ENOEXEC, so bash falls back to running the script itself. It works but it's surprising and could trip up during maintenance.

**Review by gilles-peskine-arm — APPROVED** [review]
There's a bit of weirdness that I'd like to clean up, but the code works, so I'm approving this pull request in the interest of getting this in. I would like a follow-up to clean up the weirdness.
